Automattic/woocommerce-payments
# Pass order currency to wc_price() when not in customer's session.

Fixes #1200

#### Changes proposed in this Pull Request

* Passing order currency to `wc_price()`.
* Get intent currency from the response we get from the server.
* Save order intent currency as a fallback in case we cannot get the transaction currency from the server (for order note created from within catch block, for example).
* Handle zero decimal currency in failed refund webhook handler.

#### Testing instructions

1. Testing `capture_charge()`. Steps:
- Admin > Payments > Settings > Make sure `Issue an authorization on checkout, and capture later.` is checked.
- Install plugin that can change customer currency, like [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/).
- Setup currency plugin to the point where you can see non-USD currency from the front store, for example, for [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/), follow their [docs](https://docs.woocommerce.com/document/multi-currency) and open front store with parameter `?currency=EUR` to see prices in Euro (if you setup Euro as one of the customer currencies).
- Buy a product with certain credit card numbers listed below for each test case:
  - Success: 4242424242424242 with any expired date in the future and any 3-digit CVC code.
  - Error: (I cannot find suitable testing card number that can get us to the 2 possible non-success cases [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R1161)] [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R1172)], it has to be one that passes auth step but fails capture).
- Open the newly created order from admin.
- Select 'Capture charge' from the Order actions dropdown.
- Click the 'Update' button.
- With this branch, expect order note shows amount with correct order currency. With base branch, order note will use store currency.
- Test more cases where you change the order currency from the admin order page, for example [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/) can do that. Go to the very bottom of the order admin page and you will see a dropdown to change the order currency. Make sure to click the 'Update' button at the top right. After changing the order currency, go through the above steps again and make sure order note still uses the currency when the order was placed.

2. Testing `process_refund()`. Steps:
- Install plugin that can change customer currency, like [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/).
- Setup currency plugin to the point where you can see non-USD currency from the front store, for example, for [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/), follow their [docs](https://docs.woocommerce.com/document/multi-currency) and open front store with parameter `?currency=EUR` to see prices in Euro (if you setup Euro as one of the customer currencies).
- Buy a product with certain credit card numbers listed below for each test case:
  - Success refund: 4242424242424242 with any expired date in the future and any 3-digit CVC code.
  - Failed refund: 4000000000000259 with any expired date in the future and any 3-digit CVC code. This is to test the failed case. You will see an empty alert (filed already as [separate issue](https://github.com/Automattic/woocommerce-payments/issues/1203)). Refresh the page to see the order note. That should be to test the [failed case](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R789).
- Open the newly created order from admin.
- Issue a refund, without and with reason, to test success cases respectively [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R804)] [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R810)].
- With this branch, expect order note shows amount with correct order currency. With base branch, order note will use store currency.
- Test more cases where you change the order currency from the admin order page, for example [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/) can do that. Go to the very bottom of the order admin page and you will see a dropdown to change the order currency. Make sure to click the 'Update' button at the top right. After changing the order currency, go through the above steps again and make sure order note still uses the currency when the order was placed.

3. Testing [refund failure webhook handler](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-cea8dab22a014dc0e668f8fd148aaeecd2863b3eb09cbe1f9408092c9b8d6391R194) case.
- Purchase with testing card 4000000000005126 to make refund fail.
- Test with zero decimal currency like JPY to make sure amount is correct.
<img width="289" alt="Screen Shot 2021-02-01 at 09 19 53" src="https://user-images.githubusercontent.com/73803630/106407498-c937f580-646e-11eb-82fd-6ced8ca8fdfd.png">

4. Testing [subscription renewal hook](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-a5e8d94b5af9b7536f23cf436b4eef1fb76da3069f34a63b4cf35c46051e417bR155).
I am not sure how to set this up. I just tested it with [unit test](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-933977839629fb49196cf683653de08b29c2f319c0cc0edb91ecac15d1e747a6R256-R282).

-------------------
